x/staking/keeper/msg_server.go

```diff
@@ -489,3 +492,111 @@ func (k msgServer) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParam
 
 	return &types.MsgUpdateParamsResponse{}, nil
 }
+
+func (k msgServer) RotateConsPubKey(goCtx context.Context, msg *types.MsgRotateConsPubKey) (res *types.MsgRotateConsPubKeyResponse, err error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	// Implementation is based on this ADR
+	// https://docs.cosmos.network/main/architecture/adr-016-validator-consensus-key-rotation
+
+	pk, ok := msg.NewPubkey.GetCachedValue().(cryptotypes.PubKey)
+	if !ok {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidType, "expecting cryptotypes.PubKey, got %T", pk)
+	}
+
+	consAddr := sdk.ConsAddress(pk.Address())
+
+	// checks if NewPubKey is not duplicated on ValidatorsByConsAddr
+	validator := k.Keeper.ValidatorByConsAddr(ctx, consAddr)
+	if validator != nil {
+		return nil, types.ErrConsensusPubKeyAlreadyUsedForAValidator
+	}
+
+	valAddr, err := sdk.ValAddressFromBech32(msg.ValidatorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	validator, found := k.Keeper.GetValidator(ctx, valAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+
+	// checks if the validator is exceeding parameter MaxConsPubKeyRotations within the
+	// unbonding period by iterating ConsPubKeyRotationHistory
+	//
+	// this index will contain the keys of how many
+	isExceedingLimit, rotationsMade := k.CheckLimitOfMaxRotationsExceed(ctx, valAddr)
+	if isExceedingLimit {
+		return nil, types.ErrExceedingMaxConsPubKeyRotations
+	}
+
+	// checks if the signing account has enough balance to pay KeyRotationFee
+	// pays KeyRotationFee to community fund
+	rotationFee := k.getRotationFee(ctx, validator.GetBondedTokens(), rotationsMade)
+	sender := sdk.AccAddress(valAddr)
+
+	err = k.Keeper.bankKeeper.SendCoinsFromAccountToModule(ctx, sender, distrtypes.ModuleName, sdk.NewCoins(rotationFee))
+	if err != nil {
+		return nil, err
+	}
+
+	if sdk.AccAddress(validator.GetOperator()).String() != sender.String() {
```

Comment:
move this higher to have a earlier exit if the sender doesnt match

Reply:
this is not needed now earlier it was different code, now we are converting the valAddr to sender(which will be same)